### PR TITLE
264 animate a player

### PR DESCRIPTION
**Pull request by osvegn**
## New features
- Use animation system and components to animate the player.
- Load new texture for the player animation.
- Create new textureName enumerator.
- Use animationComponent and AnimationFrequencyComponent instead of textureName on player to animate him.
## Updated features
- Use an index to store actual position during the animation in AnimationComponent.
- Update AnimationFrequencyComponent documentation.
- Change default value for the AnimationFrequencyComponent frequency.
## Fixed features
- Fix crash in AnimationSystem

### CODE REVIEW

File: libs/GraphicECS/SFML/Systems/AnimationSystem.cpp [hunk 1]
```diff
@@ -22,23 +22,24 @@ void AnimationSystem::run(World &world)
...
         world.joinEntities<GraphicsRectangleComponent, AnimationComponent, AnimationFrequencyComponent>();
 
-    auto shape = [&world](std::shared_ptr<Entity> entity) {
+    if (shapes.empty())
+        return;
+    for (auto entity : shapes) {
         using texturesNamesVector = std::vector<GraphicsTextureResource::textureName_e>;
         using texturesMap = std::unordered_map<GraphicsTextureResource::textureName_e, std::shared_ptr<sf::Texture>>;
         std::lock_guard(*entity.get());
         entity->getComponent<AnimationFrequencyComponent>().frequency -= std::chrono::duration<double>(world.getResource<GameClock>().getElapsedTime());
```
Comment (ZQUEMA) [inline libs/GraphicECS/SFML/Systems/AnimationSystem.cpp:31]: You must lock every resources

File: libs/GraphicECS/SFML/Systems/AnimationSystem.cpp [hunk 2]
```diff
@@ -22,23 +22,24 @@ void AnimationSystem::run(World &world)
...
         if (entity->getComponent<AnimationFrequencyComponent>().frequency < std::chrono::duration<double>(0)) {
             texturesNamesVector texturesNames = entity->getComponent<AnimationComponent>().textures;
-            GraphicsTextureResource::textureName_e &currentTexture =
-                entity->getComponent<AnimationComponent>().currentTexture;
+            std::size_t &currentTexturePos =
+                entity->getComponent<AnimationComponent>().currentTexturePos;
             texturesMap textures = world.getResource<GraphicsTextureResource>()._texturesList;
```
Comment (ZQUEMA) [inline libs/GraphicECS/SFML/Systems/AnimationSystem.cpp:36]: You must lock every resources

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
Lock every resources and it will be good for me

File: libs/GraphicECS/SFML/Systems/AnimationSystem.cpp [hunk 3]
```diff
@@ -22,23 +22,30 @@ void AnimationSystem::run(World &world)
...
+    if (shapes.empty())
+        return;
+    for (auto entity : shapes) {
         using texturesNamesVector = std::vector<GraphicsTextureResource::textureName_e>;
         using texturesMap = std::unordered_map<GraphicsTextureResource::textureName_e, std::shared_ptr<sf::Texture>>;
-        std::lock_guard(*entity.get());
-        entity->getComponent<AnimationFrequencyComponent>().frequency -= std::chrono::duration<double>(world.getResource<GameClock>().getElapsedTime());
+        auto guard = std::lock_guard(*entity.get());
+        {
+            auto guard = std::lock_guard(world.getResource<GameClock>());
```
Comment (ZQUEMA) [inline libs/GraphicECS/SFML/Systems/AnimationSystem.cpp:32]: Please, put the resource in a variable. That's one less action

File: libs/GraphicECS/SFML/Systems/AnimationSystem.cpp [hunk 4]
```diff
@@ -22,23 +22,30 @@ void AnimationSystem::run(World &world)
...
+                std::chrono::duration<double>(world.getResource<GameClock>().getElapsedTime());
+        }
         if (entity->getComponent<AnimationFrequencyComponent>().frequency < std::chrono::duration<double>(0)) {
             texturesNamesVector texturesNames = entity->getComponent<AnimationComponent>().textures;
-            GraphicsTextureResource::textureName_e &currentTexture =
-                entity->getComponent<AnimationComponent>().currentTexture;
-            texturesMap textures = world.getResource<GraphicsTextureResource>()._texturesList;
+            std::size_t &currentTexturePos = entity->getComponent<AnimationComponent>().currentTexturePos;
+            {
+                auto guard = std::lock_guard(world.getResource<GraphicsTextureResource>());
```
Comment (ZQUEMA) [inline libs/GraphicECS/SFML/Systems/AnimationSystem.cpp:40]: Please, put the resource in a variable. That's one less action

File: libs/GraphicECS/SFML/Systems/AnimationSystem.cpp (hunk 3 above)
Comment (ZQUEMA) [inline libs/GraphicECS/SFML/Systems/AnimationSystem.cpp:32]: rename the variable `auto guard`. it has the same name as the one just above.

File: libs/GraphicECS/SFML/Systems/AnimationSystem.cpp (hunk 4 above)
Comment (ZQUEMA) [inline libs/GraphicECS/SFML/Systems/AnimationSystem.cpp:40]: rename the variable auto guard. it has the same name as the others  just above.

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
Some other changes are required

File: libs/GraphicECS/SFML/Systems/DrawComponents.cpp [hunk 5]
```diff
@@ -63,8 +70,19 @@ void DrawComponents::run(World &world)
 
                 entityPtr->addComponent<GraphicsRectangleComponent>(
                     entityPos.x, entityPos.y, entitySize.x, entitySize.y);
-                if (layerType.layer == LayerLvL::layer_e::PLAYER)
-                    entityPtr->addComponent<TextureName>(GraphicsTextureResource::PLAYER_STATIC);
+                if (layerType.layer == LayerLvL::layer_e::PLAYER) {
```
Comment (31Nathan) [inline libs/GraphicECS/SFML/Systems/DrawComponents.cpp:73]: Please Make a function

**Review by 31Nathan — CHANGES_REQUESTED** [review]

**Review by ZQUEMA — APPROVED** [review]

File: libs/GraphicECS/SFML/Systems/DrawComponents.cpp (hunk 5 above)
Comment (osvegn) [reply libs/GraphicECS/SFML/Systems/DrawComponents.cpp:73]: It will be improved in #270

**Review by TomDUVAL-MAHE — APPROVED** [review]

**Review by 31Nathan — DISMISSED** [review]
Do not forget to make functions ^^

**Review by TomDUVAL-MAHE — APPROVED** [review]

**Review by 31Nathan — APPROVED** [review]
LGTM

**Pull request merged**